### PR TITLE
Homework - Cars

### PR DESCRIPTION
**Pull request by MaciejSalamonski**
- exceptions for InvalidGear, InvalidSpeed and InvalidEngineChange
- encapsulation, class members are private or protected
- add variable speed_, private  member to Car class
- brake() set speed_ to zero.
- in function accelerate() we check that the speed is not a negative value
- some setters and getters added
- function carInfo() shows info about our car
- the usnigned types were used. This protects the program in a basic way from bad input. Examples: power, gears
- check for proper engine type in changeEngine
- check if car velocity is zero to changeEngine call
- changing validation: We can only shift to reverse gear when we are in neutral gear. Also we cant shift gear to one that does not exist.
- deleting engines in destructors and after engine change.  
- tests added
- workflow file added

### CODE REVIEW

File: Car.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void Car::brake() {
+    std::cout << __FUNCTION__ << std::endl;
+    setSpeed(MINIMUM_VELOCITY);
+}
+
+void Car::accelerate(int speed) {
+    if (speed < MINIMUM_VELOCITY) {
```
Comment (WBigaj) [inline Car.cpp:29]: What flags are you using during compiling? You are comparing here int with bool. I would be suprised if compilator with all flags would not generate warning here.

File: HybridCar.cpp [hunk 2]
```diff
@@ -1,19 +1,58 @@
...
+    }
+
+    if (typeid(*engine) == typeid(PetrolEngine)) {
+        delete petrolEngine_;
+        petrolEngine_ = static_cast<PetrolEngine*>(engine);
+        std::cout << __FUNCTION__ << std::endl;
+        return;
+    }
 
+    return;
```
Comment (WBigaj) [inline HybridCar.cpp:57]:
```suggestion
```
It's not needed here since void does not need to return anything.

File: HybridCar.cpp [hunk 3]
```diff
@@ -1,19 +1,58 @@
...
+        electricEngine_ = static_cast<ElectricEngine*>(engine);
+        std::cout << __FUNCTION__ << std::endl;
+        return;
+    }
+
+    if (typeid(*engine) == typeid(PetrolEngine)) {
+        delete petrolEngine_;
+        petrolEngine_ = static_cast<PetrolEngine*>(engine);
+        std::cout << __FUNCTION__ << std::endl;
+        return;
```
Comment (WBigaj) [inline HybridCar.cpp:54]:
```suggestion
```

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: Car.cpp (hunk 1 above)
Comment (MaciejSalamonski) [reply Car.cpp:29]: Honestly, this variable was supposed to be uint8_t xd. Im using g++ 9.3.0

File: HybridCar.cpp (hunk 2 above)
Comment (MaciejSalamonski) [reply HybridCar.cpp:57]: Thanks i forgot to delete this

File: Car.cpp (hunk 1 above)
Comment (WBigaj) [reply Car.cpp:29]: I was asking about flags(like -wall -pedantic). It should warn you that you are comparing different types.

**Review by Kasperice — APPROVED** [review]
Good job, I don't have any comments

File: ElectricCar.cpp [hunk 4]
```diff
@@ -1,16 +1,50 @@
...
+              << "Power: " << getEnginePower() << '\n'
+              << "Battery Capacity: " << getBatteryCapacity() << '\n';
+}
+
+uint16_t ElectricCar::getEnginePower() const {
+    return electricEngine_->getPower();
+}
+
+void ElectricCar::changeEngine(Engine* engine) {
+    if (getSpeed() != MINIMUM_VELOCITY) {
```
Comment (ziobron) [inline ElectricCar.cpp:40]: unsafe operation

**Review by ziobron — CHANGES_REQUESTED** [review]
Please correct a type of MINIMUM_VELOCITY

File: Car.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "Engine.hpp"
+
+constexpr bool MINIMUM_VELOCITY = 0;
```
[the remark this answers is not in the document]
Comment (MaciejSalamonski) [reply Car.hpp]: Done

File: ElectricCar.cpp (hunk 4 above)
Comment (MaciejSalamonski) [reply ElectricCar.cpp:40]: Could you explain what is unsafe there? What should i change?

Comment (WBigaj) [reply ElectricCar.cpp:40]: Now it's safe since you are comparing int with int.

Comment (MaciejSalamonski) [reply ElectricCar.cpp:40]: Ohh Ok. Thanks @WBigaj ;~}

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
8 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**